R-package/R/xgboost.R

```diff
@@ -1,32 +1,1022 @@
+    })
+    y <- as.data.frame(y) |> as.matrix()
+
+    if (!is.null(base_margin)) {
+      base_margin <- process.base.margin(base_margin, length(y), ncol(y))
+    }
+
+    out <- list(
+      params = list(
+        objective = objective
+      ),
+      dmatrix_args = list(
+        label = y,
+        base_margin = base_margin
+      ),
+      metadata = list(
+        y_names = y_names,
+        n_targets = ncol(y)
+      )
+    )
+
+  } else if (is.matrix(y)) {
+    if (ncol(y) == 1L) {
+      return(process.y.margin.and.objective(as.vector(y), base_margin, objective, params))
+    }
+
+    if (!is.null(objective) && !(objective %in% .MULTI_TARGET_OBJECTIVES())) {
+      stop(
+        "Got multi-column 'y' - supported objectives for this data are: ",
+        paste(.MULTI_TARGET_OBJECTIVES(), collapse = ", "),
+        ". Was passed: ",
+        objective
+      )
+    }
+    if (is.null(objective)) {
+      objective <- "reg:squarederror"
+    }
+
+    y_names <- colnames(y)
+    if (storage.mode(y) != "double") {
+      storage.mode(y) <- "double"
+    }
+
+    if (!is.null(base_margin)) {
+      base_margin <- process.base.margin(base_margin, nrow(y), ncol(y))
+    }
+
+    out <- list(
+      params = list(
+        objective = objective
+      ),
+      dmatrix_args = list(
+        label = y,
+        base_margin = base_margin
+      ),
+      metadata = list(
+        n_targets = ncol(y)
+      )
+    )
+
+    if (NROW(y_names) == ncol(y)) {
+      out$metadata$y_names <- y_names
+    }
+
+  } else {
+    stop("Passed 'y' object with unsupported class: ", paste(class(y), collapse = ", "))
+  }
+
+  return(out)
+}
+
+process.row.weights <- function(w, lst_args) {
+  if (!is.null(w)) {
+    if ("label" %in% names(lst_args$dmatrix_args)) {
+      nrow_y <- NROW(lst_args$dmatrix_args$label)
+    } else if ("label_lower_bound" %in% names(lst_args$dmatrix_args)) {
+      nrow_y <- length(lst_args$dmatrix_args$label_lower_bound)
+    } else {
+      stop("Internal error.")
+    }
+    if (!is.numeric(w)) {
+      w <- as.numeric(w)
+    }
+    if (length(w) != nrow_y) {
+      stop(
+        "'weights' must be a 1-d vector with the same length as 'y' (",
+        length(w), " vs. ", nrow_y, ")."
+      )
+    }
+    lst_args$dmatrix_args$weight <- w
+  }
+  return(lst_args)
+}
+
+check.nthreads <- function(nthreads) {
+  if (is.null(nthreads)) {
+    return(1L)
+  }
+  if (!inherits(nthreads, c("numeric", "integer")) || !NROW(nthreads)) {
+    stop("'nthreads' must be a positive scalar value.")
+  }
+  if (length(nthreads) > 1L) {
+    nthreads <- head(nthreads, 1L)
+  }
+  if (is.na(nthreads) || nthreads < 1) {
+    stop("Passed invalid 'nthreads': ", nthreads)
+  }
+  if (is.numeric(nthreads)) {
+    if (floor(nthreads) != nthreads) {
+      stop("'nthreads' must be an integer.")
+    }
+  }
+  return(as.integer(nthreads))
+}
+
+check.can.use.qdm <- function(x, params) {
+  if (inherits(x, "sparseMatrix") && !inherits(x, "dgRMatrix")) {
```

Comment:
We can support COO internally, for CSC, maybe we can simply transpose in R. Do you think it can help the make code less complex?

Reply:
Maybe, but what would be the benefit in such case if the data still needs to be duplicated?

Reply:
Modified it to accept COO and CSC for QDM, with a casting for now until direct QDM construction from COO is added.